### DISCUSSION TITLE
Performance in analysis and generation

### DISCUSSION THREAD

**Discussion by ryo33** [jsinger67/parol#29] [Ideas]
I've only seen a tiny portion of the code, and it's mostly my impression of how it works in my project, `parol` does not seem to focus primarily on speed in analysis and generation.
The culture of Rust has been that slow compilation is acceptable and unavoidable, but recently there have been changes such as incremental compile support in `rustc`, faster linkers, and the emergence of libraries that sell quick compilation for rapid development iteration, e.g. [Bevy](https://bevyengine.org).
I don't think we don't need exhaustive optimization like written in [this](https://gist.github.com/kvark/f067ba974446f7c5ce5bd544fe370186), and I think we only need faster data-structures/algorithms and parallelization in hot-spots. I love simple codes, and I think even more so.
Do you have any plans to improve the compilation speed? (I've noticed you've starred https://github.com/rust-lang/measureme.)
I have no practical experience in profiling, but I know some best practices in Rust, both profiling and optimization. I wish I could help.

## Related links

- [cargo flamegraph](https://github.com/flamegraph-rs/flamegraph)
- [tracing](https://docs.rs/tracing/latest/tracing/)

**Comment by jsinger67** [comment]
In the early phases of this crate I spent some effort in optimizing the performance of analysis. Mostly in the parts where FOLLOW(k) is calculated, especially when k gets bigger.
I did these analyses mostly with the help of Visual Studio under Windows which is my primary development platform.
I also digged a bit into possible parallelization there but gave it up because of a lack of time and the need to bring more essential features.
I really could use a little help here and would be glad when you can jump in ☺️
Performance improvement is always good and one can put much time and effort into it.

**Reply by ryo33** [reply]
Thanks for your comment. I will try to improve the performance as my side project.
